### ISSUE TITLE
Failed to find function to patch

### ISSUE THREAD

**Issue by petsoukos** [djrbliss/loki#87]
Hi there,

I'm trying out your tool on my LG G2 device. It is an international version d802 with the bootloader locked. When I try your tool I get the message "Failed to find function to patch."

The device is rooted and I did all the steps up to the point where I run the loki_tool patch. (Which for some reason doesn't run on my Linux machine and I run it on the device itself.)

I also tested if the device didn't need to be patched, by replacing the recovery directly, boot it fails to enter the recovery because of the locked bootloader.

Any help or advice would be much appreciated.

**Comment by tsujamin** [comment]
coincidentally I encountered the same problem today too. D802T running Optus's stock 5.0.2 ROM.

I've dumped aboot.img onto my server: http://service.bgroberts.id.au/aboot.img

**Comment by tsujamin** [comment]
If you flash a plain recovery image, the following is printed 
````
ERROR : boot certification verify
(other things)
```

I've been playing with the dissassembly of the aboot image. The code starts at 0x24 and at base 0xf800000. I think the following excerpt is roughly where image_verify is called, but I'm not confident enough at disassembly, and short an IDA pro license, to confirm this.

```
loc_f818c60:
    sub_f81d8dc("ERROR : boot certification verify\n"); # First message following verification failure
    r1 = "boot certification verify";
    sub_f823f4c(0xf853050, r1);
    r0 = loc_f8173d8();
    goto loc_f818cc8;
    
loc_f818c34:
    lr = 0x800;
    r5 = *(r6 + 0x20);
    r4 = *(r6 + 0x24);
    var_C = r12;
    var_10 = r5;
    var_8 = sp + 0x3c;
    if (loc_f82411c() == 0x0) goto loc_f818bc8; # is loc_f82411c() image_verify perhaps?
    goto loc_f818c60;
```

**Comment by petsoukos** [comment]
Were you able to run the loki_tool on your machine or do we have to run it on the device itself? Because the instructions start with "dan@pc" and not at "shell@android"... maybe I'm missing something.

**Comment by tsujamin** [comment]
It won't matter. loki_tool patches the image the same regardless of where it's run so long as you give it the same aboot.img I think.

**Comment by tsujamin** [comment]
Looks like the header is loaded at `0xf8ff000` by the memcmp on the magic.
```
loc_f8193fc:
    r1 = "ANDROID!";
    r0 = memcmp(0xf8ff000, r1, 0x8);
    if (r0 == 0x0) goto loc_f8197cc;
```

Kernel/ramdisk info look _reasonable_
```
[+] Original kernel address: 00008000
[+] Original ramdisk address: 05000000
```

**Comment by petsoukos** [comment]
Ok, then. So my aboot.img isn't working with this tool. Here is my aboot.img Maybe you can take a look and better understand what's wrong: https://www.dropbox.com/s/gfkyr8tlrn10tnm/aboot.img?dl=0

**Comment by tsujamin** [comment]
Yours has a different checksum to mine (which I'm trying to get working atm), but it looks roughly the same inside, so hopefully if I can get something working it'll help both of us.

Out of curiousity, what Android Version/Build Number/Model Number does your phone report? Mine is 5.0.2/LRX22G/LG-D802T Rev 1.0

**Comment by petsoukos** [comment]
5.0.2
M8974A-AAAANAZM-1.0.190128
Kernel: 3.4.0
LRX22G
V30d-EUR-XX

**Comment by tsujamin** [comment]
Neat, they must have rebuild aboot with the 5.0.2 release

**Comment by tsujamin** [comment]
Yeah no luck on my end. I'm fairly sure hdr should be 0xf8f41c0
```
loc_f819488:
    if (memcmp(0xf8f41c0, "ANDROID!", 0x8) != 0x0) goto loc_f819ed0;
```

Can't quite work out where check_sigs is though, I've tried a handful of places. Hopefully someone with more experience can track down the rest :)

**Comment by djrbliss** [comment]
For future reference, you can use "make host" to build an x86 version of Loki for use on hosts (rather than the default, which builds ARM for device use).

Both aboot.img files provided have been patched against the vulnerability that Loki exploits, so they cannot be supported. This shouldn't be a surprise, since Loki was released ~2.5 years ago. Closing.

**Comment by staminna** [comment]
I tried to make host but

ld: library not found for -lcrt0.o
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [loki_tool] Error 1
Jorges-MBP:bts jorge$ make host -v
GNU Make 3.81
